### PR TITLE
[EDA] Prevent Deletion of Plan Requirement Record  if There are Associated Child Records

### LINKED ISSUE #1125: [EDA] Prevent Deletion of Plan Requirement Record  if There are Associated Child Records
**Note: FOR INTERNAL USE ONLY. For support questions, please use the EDA Chatter group in [The Power of Us Hub](https://powerofus.force.com).
For feature requests, please use the Ideas tab in [The Power of Us Hub](https://powerofus.force.com).**

* **Do you want to request a *feature* or report a *bug*?**
Feature
* **What is the current behavior?**
When a parent Plan Requirement with a child related Plan Requirement is deleted, the user is unable to delete the Plan Requirement unless the user disables PReq_CannotDelete_TDTM via trigger handler record. 
* **What is the expected behavior?**
A user should be able to control whether they are able to delete a parent Plan Requirement with a child related Plan Requirement is deleted, by enabling/disabling this feature via EDA settings/hierarchy custom settings instead of disabling PReq_CannotDelete_TDTM trigger handler record. 
* **What is the motivation / use case for changing the behavior?**
data integrity

### PR DESCRIPTION
**Pull request by kdy618**
# Critical Changes

# Changes
We've added functionality to prevent the deletion of Plan Requirement records that have related child records. Plan Requirement records with child records other than the child Plan Requirement can be deleted. This change is for internal-use only and is enabled at this time. If customers disable PReq_CannotDelete_TDTM via trigger handler record, regardless of whether this feature is enabled via Hierarchy Settings, this feature will not prevent deletion of Plan Requirements and vice versa.

Note: We strongly recommend that EDA Admins don't manage Hierarchy Custom Settings directly and, instead, use the EDA Settings UI. 
# Issues Closed
Closes #1125
# New Metadata

# Deleted Metadata

# Testing Notes
https://salesforce.quip.com/77mPAdh0IjGQ#QSNACAut9Lq

### CODE REVIEW

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 1]
```diff
@@ -35,30 +35,39 @@
...
+  * @return dmlWrapper.
+  ********************************************************************************************************/
+  public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+                                 TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+    
+      if (!enabledPreventPlanRequirementDeletion) {
+        return new DmlWrapper(); 
+      }
+                                     
+      Map<ID, Plan_Requirement__c> oldmap = new Map<ID, Plan_Requirement__c>((List<Plan_Requirement__c>)oldlist);
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]: @kdy618 This is currently using two spaces for indentations.

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 2]
```diff
@@ -35,30 +35,39 @@
...
+                                              FROM Plan_Requirement__c
+                                              WHERE ID IN :oldlist]) {
+              if(planReq.Plan_Requirements__r.size() > 0 ) {
+                  Plan_Requirement__c planReqInContext = oldmap.get(planReq.ID);
+                  planReqInContext.addError(Label.CannotDelete);
+              }
+          }
+      }
+      return new DmlWrapper();
+  }
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]: This also is using two spaces for indentations.

File: src/classes/PReq_CannotDelete_TEST.cls [hunk 3]
```diff
@@ -36,33 +36,111 @@
 */
 @isTest
 public with sharing class PReq_CannotDelete_TEST {
-
+  /*********************************************************************************************************
+    * @description Retrieves the Business Account record type Id. 
+    */
+    public static String bizAccRecordTypeId = UTIL_Describe_API.getBizAccRecTypeID();
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TEST.cls]: Given the underlying code in UTIL_Describe for this, I want to flag this for potential cleanup when we address that issue. The desired result here seems to be to get an account record type that does not have existing TDTM behavior being run against it.

File: src/classes/PReq_CannotDelete_TEST.cls [hunk 4]
```diff
@@ -36,33 +36,111 @@
 */
 @isTest
 public with sharing class PReq_CannotDelete_TEST {
-
+  /*********************************************************************************************************
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TEST.cls]: This is using two rather than four spaces.

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 5]
```diff
@@ -35,30 +35,39 @@
...
+  
+  /*******************************************************************************************************
+  * @description Stops an Plan Requirement from being deleted if it has any child Plan Requirement
+  * @param listNew the list of Plan Requirement from trigger new.
+  * @param listOld the list of Plan Requirement from trigger old.
+  * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.).
+  * @param objResult the describe for Plan Requirement
+  * @return dmlWrapper.
+  ********************************************************************************************************/
+  public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist,
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]: These present an opportunity for us to split this out into separate methods to assist with refactoring and focused unit tests. In the future, I'd like us to consider minimizing the amount of logic we have in run methods.

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 6]
```diff
@@ -35,30 +35,48 @@
...
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]:
```suggestion
    /*******************************************************************************************************
    * @description Get the setting of preventing Account deletion
    */
    private static Boolean enabledPreventPlanRequirementDeletion = UTIL_CustomSettingsFacade.getSettings().Prevent_Plan_Requirement_Deletion__c;
  
    /*******************************************************************************************************
    * @description Stops an Plan Requirement from being deleted if it has any child Plan Requirement
    * @param listNew the list of Plan Requirement from trigger new.
    * @param listOld the list of Plan Requirement from trigger old.
    * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.).
    * @param objResult the describe for Plan Requirement
    * @return dmlWrapper.
    ********************************************************************************************************/
    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
```

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 7]
```diff
@@ -35,30 +35,48 @@
...
+  * @return dmlWrapper.
+  ********************************************************************************************************/
+  public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+                                 TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+    
+      if (!enabledPreventPlanRequirementDeletion) {
+        return new DmlWrapper(); 
+      }
+                                     
+      Map<ID, Plan_Requirement__c> oldmap = new Map<ID, Plan_Requirement__c>((List<Plan_Requirement__c>)oldlist);
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]:
```suggestion
                                 TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
    
        if (!enabledPreventPlanRequirementDeletion) {
            return new DmlWrapper(); 
        }
                                     
        Map<ID, Plan_Requirement__c> oldmap = new Map<ID, Plan_Requirement__c>((List<Plan_Requirement__c>)oldlist);
```

Comment (MatthewBlanski) [reply src/classes/PReq_CannotDelete_TDTM.cls]: Additional spacing changes.

File: src/classes/PReq_CannotDelete_TDTM.cls [hunk 8]
```diff
@@ -35,30 +35,48 @@
...
+      return new DmlWrapper();
+  }
+  
+   /*******************************************************************************************************
+   * @description Evaluates whether the Plan Requirement has any child related records.
+   * @param pr is the current Plan Requirement record.
+   * @return Boolean.
+   ********************************************************************************************************/
+    private static Boolean hasChildRecords(Plan_Requirement__c pr) {
+        return (pr.Plan_Requirements__r.size() > 0);
```
Comment (MatthewBlanski) [inline src/classes/PReq_CannotDelete_TDTM.cls]:
```suggestion
        if (triggerAction == TDTM_Runnable.Action.BeforeDelete) {
            for (Plan_Requirement__c planReq : [SELECT ID,
                                               (SELECT ID FROM Plan_Requirements__r LIMIT 1)
                                              FROM Plan_Requirement__c
                                              WHERE ID IN :oldlist]) {
                if(hasChildRecords(planReq)) {
                    Plan_Requirement__c planReqInContext = oldmap.get(planReq.ID);
                    planReqInContext.addError(Label.CannotDelete);
                }
            }
        }
        return new DmlWrapper();
      }
  
    /*******************************************************************************************************
    * @description Evaluates whether the Plan Requirement has any child related records.
    * @param pr is the current Plan Requirement record.
    * @return Boolean.
    ********************************************************************************************************/
    private static Boolean hasChildRecords(Plan_Requirement__c pr) {
        return (pr.Plan_Requirements__r.size() > 0);
```

Comment (MatthewBlanski) [reply src/classes/PReq_CannotDelete_TDTM.cls]: The last round of spacing changes

**Comment by spinkelman** [comment]
Passes QA - Merging
Will close WI

**Pull request merged**